domain/secret/service/service.go

```diff
@@ -478,17 +486,103 @@ func (s *SecretService) UpdateCharmSecret(ctx context.Context, uri *secrets.URI,
 			}
 		}()
 	}
-
-	err := s.secretState.UpdateSecret(ctx, uri, p)
-	if errors.Is(err, secreterrors.SecretLabelAlreadyExists) {
-		return errors.Errorf("secret with label %q is already being used", *params.Label)
-	}
+	err := s.secretState.RunAtomic(ctx, func(ctx domain.AtomicContext) error {
+		return s.updateSecret(ctx, uri, p)
+	})
 	if err != nil {
 		return errors.Annotatef(err, "cannot update charm secret %q", uri.ID)
 	}
 	return nil
 }
 
+func (s *SecretService) createSecret(
+	ctx domain.AtomicContext, version int, uri *secrets.URI, owner secrets.Owner, params domainsecret.UpsertSecretParams,
+) (err error) {
+	defer func() {
+		if err != nil {
+			if errors.Is(err, secreterrors.SecretLabelAlreadyExists) {
+				err = fmt.Errorf("secret with label %q is already being used: %w", *params.Label, secreterrors.SecretLabelAlreadyExists)
+			}
+		}
+	}()
+	switch kind := owner.Kind; kind {
+	case secrets.ApplicationOwner:
+		appUUID, err := s.secretState.GetApplicationUUID(ctx, owner.ID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if params.Label != nil && *params.Label != "" {
+			labelExists, err := s.secretState.CheckApplicationSecretLabelExists(ctx, appUUID, *params.Label)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if labelExists {
+				return fmt.Errorf("secret with label %q is already being used: %w", *params.Label, secreterrors.SecretLabelAlreadyExists)
+			}
+		}
+		return s.secretState.CreateCharmApplicationSecret(ctx, version, uri, appUUID, params)
+	case secrets.UnitOwner:
+		unitUUID, err := s.secretState.GetUnitUUID(ctx, owner.ID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if params.Label != nil && *params.Label != "" {
+			labelExists, err := s.secretState.CheckUnitSecretLabelExists(ctx, unitUUID, *params.Label)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if labelExists {
+				return fmt.Errorf("secret with label %q is already being used: %w", *params.Label, secreterrors.SecretLabelAlreadyExists)
+			}
```

Comment:
We can use the same style as for updateSecret()
Get the labelExists and err and then process at the end

eg

```
var (
   labelExists bool
   labelErr error
)
switch kind := owner.Kind; kind {
	case secrets.UnitOwner:
unitUUID, err := s.secretState.GetUnitUUID(ctx, owner.ID)
	if err == nil {
		return errors.Trace(err)
	}
	if params.Label != nil && *params.Label != "" {
		labelExists, labelErr = s.secretState.CheckUnitSecretLabelExists(ctx, unitUUID, *params.Label)
	}
...
}
if labelErr != nil {
	return errors.Trace(labelErr)
}
if labelExists {
	return fmt.Errorf("secret with label %q is already being used: %w", *params.Label, secreterrors.SecretLabelAlreadyExists)
}
```

Reply:
You see here, the create state methods are different as well.